src/v/raft/tests/basic_raft_fixture_test.cc

```diff
@@ -862,3 +862,101 @@ TEST_F_CORO(raft_fixture, test_no_stepdown_on_append_entries_timeout) {
     ASSERT_EQ_CORO(term_before, new_leader_node.raft()->term());
     ASSERT_TRUE_CORO(new_leader_node.raft()->is_leader());
 }
+
+/**
+ * This synthetic test is there to trigger a situation in which follower
+ * receives an append entries request which contains only batches that matches
+ * its log. This trigger a condition in which the follower should reply with
+ * success to the leader so the leader can continue recovery process.
+ *
+ * The test uses reply interception to 'trick' the leader to send the append
+ * entries with the batches that the follower already has.
+ */
+TEST_F_CORO(raft_fixture, test_redelivery_of_matching_logs) {
+    co_await create_simple_group(3);
+    auto leader_id = co_await wait_for_leader(10s);
+    model::node_id non_leader_id{0};
+    for (auto& [id, n] : nodes()) {
+        n->set_default_recovery_read_size(1);
+    }
+
+    for (auto& [id, _] : nodes()) {
+        if (id != leader_id) {
+            non_leader_id = id;
+            break;
+        }
+    }
+    auto& leader_node = node(leader_id);
+    /**
+     * Replicate data to all nodes
+     */
+    auto r = co_await leader_node.raft()->replicate(
+      make_batches(200, 1, 10),
+      replicate_options(consistency_level::quorum_ack, 10s));
+    /**
+     * Prevent one node from receiving append entries
+     */
+    leader_node.on_dispatch(
+      [non_leader_id](model::node_id id, raft::msg_type t) {
+          if (t == raft::msg_type::append_entries && id == non_leader_id) {
+              throw std::runtime_error("error");
+          }
+          return ss::now();
+      });
+    /**
+     * Replicate data with one of the nodes being blocked
+     */
+    r = co_await leader_node.raft()->replicate(
+      make_batches(200, 1, 10),
+      replicate_options(consistency_level::quorum_ack, 10s));
+    ASSERT_FALSE_CORO(r.has_error());
+
+    /**
+     * Append some batches to the blocked node with different term
```

Comment:
How can this happen naturally? This node's log is shorter than the majority, so it cannot be elected as a leader.

In my test data was produced with acks=1, isn't it what we need here to keep the scenario plausible?

Reply:
for a short while when a node is still a leader but the others doesn't receive messages it may always happen, using `acks=1` makes it more plausible

Reply:
Could you check my understanding of how we can make it happen without manual log append?
1. Elect node `not_leader` as a leader in term 1
2. Produce, do not wait for replicated, make sure (by blocking dispatch or append_entries?) it does NOT get replicated to `leader` or `third`
3. Elect node `leader` as a leader in term 2 (we'll need to make node `third` communicated well with `leader` for it)
4. Block off append entries on `not_leader`
5. Produce again, make sure it gets replicated to `third` and thus committed (is it vital for the test?)
6. Proceed with the rest of the test.

The terms would be the other way round (1 on `leader`, 2 on `not_leader`), I'm not sure how to make them the right way round.

Is that right? Do we not do this just to keep the test simple?

Reply:
exactly, the description here is valid, i just wanted to make it a bit simpler

Reply:
Could you swap round term values, so that the situation looks more realistic? Or explain how term 2 data can be appended to `non_leader`'s log.

Reply:
i added a comment and made it more clear with terms